src/test/shared/utilities/collectionUtils.test.ts

```diff
@@ -538,4 +541,127 @@ describe('CollectionUtils', async function () {
             })
         })
     })
+
+    const promise = <T>(val: T) => new Promise<T>(resolve => setImmediate(() => resolve(val)))
+    const cons = <T, U>(arr: Promise<T>[], next: U) => {
+        if (arr.length === 0) {
+            return [promise(next)]
+        } else {
+            return [...arr, arr[arr.length - 1].then(() => promise(next))]
+        }
+    }
+    const toPromiseChain = <T>(arr: T[]) => arr.reduce(cons, [] as Promise<T>[])
+
+    async function iterateAll<T>(iterable: AsyncIterable<T>) {
+        const result = [] as (T | Error)[]
+        try {
+            for await (const val of iterable) {
+                result.push(val)
+            }
+        } catch (err) {
+            assert.ok(err instanceof Error)
+            result.push(err)
+        }
+
+        return result
+    }
+
+    describe('toStream', function () {
+        it('yields values that resolve first regardless of order', async function () {
+            const values = toPromiseChain([0, 1, 2, 3])
+            const stream = toStream(values.reverse())
+
+            assert.deepStrictEqual(await iterateAll(stream), [0, 1, 2, 3])
+        })
+
+        it('immediately yields values that are not promises', async function () {
+            const stream = toStream([0, promise(1), 2, promise(3)])
+
+            assert.deepStrictEqual(await iterateAll(stream), [0, 2, 1, 3])
+        })
+
+        it('throws on any error', async function () {
```

Comment:
Exceptions and concurrent tasks is an interesting topic. I think ideally we would want to give the caller the opportunity to handle _all_ exceptions from the iterable. If the goal is not to 'stream' the results, then `allSettled` would work pretty well. But when the results are consumed one at a time, it's better to handle the errors just-in-time instead of waiting until all tasks resolve or reject. There's not a good way to do this with native JS unfortunately. But I could add a `catch` method to `AsyncCollection` that would allow callers to do the same thing as `allSettled` if they need that behavior.

Reply:
Ah right, `allSettled()` (and `all()`) doesn't make sense for a stream. `race()` is used precisely because a stream doesn't have a concept of "all".